### DISCUSSION TITLE
no way found to detect/filter nodes with embedded images that have outdated links

### DISCUSSION THREAD

**Discussion by Bricklam** [freeplane/freeplane#1087] [Help]
I have a common problem with nodes that had images embedded (linked) into them, after a while if the linked image file was moved to another location or renamed, the node no longer display the image, instead it would display this text:
```
FileNotFoundException
<image path> (The system cannot find the file specified)
```

here is an example:

![Screenshot1](https://user-images.githubusercontent.com/129968602/230806255-63058ef2-9260-4041-83a7-b37254b1338f.PNG)
![Screenshot2](https://user-images.githubusercontent.com/129968602/230806261-a26e10e5-a7c1-477e-b860-efdd657c72bd.PNG)

the problem is: how to filter/select/identify all the nodes that have (FileNotFoundException / invalid image links), in order to fix their broken image links?
I have tested many ways and all filter options, but didn't find any available way
does this task need a dedicated a script? or is there any hidden way to get these nodes?

thanks

**Comment by EdoFro** [comment]
I tried to find an easy way, but I couldn’t.

**Reply by EdoFro** [reply]
@Bricklam ,
try `Script filter`:

```groovy
def uri = node.externalObject.uri?.toURI()
if(uri && uri.scheme == 'file'){
    return !(new File(uri).exists())
} else {
    return false
}
```

Hope it helps,
edo

**Comment by abc16361** [comment]
Maybe this could help. [Check for broken links](https://docs.freeplane.org/scripting/Scripts_collection.html#check-for-broken-links)

**Comment by beccon4** [comment]
I've made the proposal into the Issues Section yesterday to save copies of the image file within the Mindmap itself - to embed them in a CDATA tag as base64 or something. That would ease the processing of Mindmaps with images in them a lot. I came across this when I prepared a presentation with illustrations and sent it off to someone else later who was subsequently unable to recover the links to put it all together again. I'm not sure whether I included all the jpegs - there is no way to quickly find that out :-) To make it perfect: allow scaling the resolution to avoid the Mindmap file getting too big. 

Please also allow to paste images into notes and html node texts. I've tried this and Freeplane got stuck for a while (Ubuntu 22.04). That would also work with html-Links to a CDATA Section ...

**Comment by ldmpub** [comment]
I was faced to this problem this week => ie missing images after a map rename + move.
Hopefully I had only few images (around 5) in my map and errors raised were displayed "so big" that I can not miss them 🤣.

After browsing the map to find each node with missing image, I used the "Replace" command to change the default image location.

I wish to have a native "broken link" or "image link checker" tool in Freeplane to identify then navigate to erroneous nodes.
Maybe next step could be to have a batch replace feature from "old folder" to "new folder" on images (to be honest I thought for a moment I will perform a search and replace of images path in the mm / XML file itself 😅).

PS for @beccon4 : As far as I have understood, to embed images in Freeplane mindmaps is not a good idea at all because it can cause a lot of performance issues ...
